### PR TITLE
feat: add secret for java sdk

### PR DESCRIPTION
**Pull request by fft0518**
The PR is add secret api for java sdk and is one of the ospp projects.

In this PR, I have a problem that I don't how to convert SecretResponse type to Map type.  I also use the SecretResponse type although I think it's wrong.

https://github.com/mosn/layotto/issues/626

### CODE REVIEW

File: sdk/src/main/java/spec/sdk/runtime/v1/domain/SecretRuntime.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package spec.sdk.runtime.v1.domain;
+
+import spec.sdk.runtime.v1.domain.secret.GetBulkSecretRequest;
+import spec.sdk.runtime.v1.domain.secret.GetBulkSecretResponse;
+import spec.sdk.runtime.v1.domain.secret.GetSecretRequest;
+import spec.sdk.runtime.v1.domain.secret.GetSecretResponse;
+
+public interface SecretRuntime {
```
Comment (seeflood) [inline sdk/src/main/java/spec/sdk/runtime/v1/domain/SecretRuntime.java:8]: `src/main/java/spec/sdk/runtime/v1/client/RuntimeClient.java` should also extend this interface
image

**Review by seeflood — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/spec/sdk/runtime/v1/domain/SecretRuntime.java (hunk 1 above)
Comment (fft0518) [reply sdk/src/main/java/spec/sdk/runtime/v1/domain/SecretRuntime.java:8]: Thanks.

**Comment by ZLBer** [comment]
image

@fft0518  it has method to get `sercretsMap`

File: examples-secret/src/main/java/io/mosn/layotto/examples/secret/Secret.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import spec.sdk.runtime.v1.domain.secret.GetBulkSecretRequest;
+import spec.sdk.runtime.v1.domain.secret.GetBulkSecretResponse;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Secret {
+
+
+
```
Comment (ZLBer) [inline examples-secret/src/main/java/io/mosn/layotto/examples/secret/Secret.java:17]: format code, remove blank line

File: sdk/src/main/java/io/mosn/layotto/v1/RuntimeClientGrpc.java [hunk 3]
```diff
@@ -1101,4 +1105,61 @@ public UnlockResponse unlock(UnlockRequest request) {
...
+                    .getSecret(request);
+            GetSecretResponse getSecretResponse = new GetSecretResponse();
+            getSecretResponse.setData(response.getDataMap());
+            return getSecretResponse;
+        } catch (Exception e) {
+            logger.error("getSecret error ", e);
+            throw new RuntimeClientException(e);
+        }
+
+
```
Comment (ZLBer) [inline sdk/src/main/java/io/mosn/layotto/v1/RuntimeClientGrpc.java:1129]: same as above

**Comment by ZLBer** [comment]
@fft0518  add a java example? like [examples-sequencer](https://github.com/layotto/java-sdk/tree/main/examples-sequencer)

**Comment by ZLBer** [comment]
@fft0518  merge the master branch before push code
image

**Comment by fft0518** [comment]
> @fft0518 add a java example? like [examples-sequencer](https://github.com/layotto/java-sdk/tree/main/examples-sequencer)
I added it at the firsttime

**Review by ZLBer — DISMISSED** [review]

**Comment by fft0518** [comment]
@seeflood please help to review it. Thanks.

File: examples-secret/pom.xml [hunk 4]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <artifactId>runtime-sdk-parent</artifactId>
+        <groupId>io.mosn.layotto</groupId>
+        <version>1.2.0-SNAPSHOT</version>
```
Comment (seeflood) [inline examples-secret/pom.xml:9]: Now it's `1.3.0-SNAPSHOT`. Please replace all the `1.2.0-SNAPSHOT` with `1.3.0-SNAPSHOT`

**Review by seeflood — CHANGES_REQUESTED** [review]

File: examples-secret/pom.xml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        <finalName>examples-secret</finalName>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-assembly-plugin</artifactId>
+                <version>3.2.0</version>
+                <configuration>
+                    <archive>
+                        <manifest>
+                            <mainClass>io.mosn.layotto.examples.sequencer.Secret</mainClass>
```
Comment (seeflood) [inline examples-secret/pom.xml:59]:
```suggestion
                            <mainClass>io.mosn.layotto.examples.secret.Secret</mainClass>
```

**Review by seeflood — CHANGES_REQUESTED** [review]

File: examples-secret/src/main/java/io/mosn/layotto/examples/secret/Secret.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+                    .build();
+
+
+            /*GetSecret*/
+            Map<String, String> meta = new HashMap<>();
+            GetSecretRequest secretreq = new GetSecretRequest();
+            secretreq.setStoreName(storeName);
+            secretreq.setKey("db-user-pass:password");
+            secretreq.setMetaData(meta);
+            System.out.println(client.getSecret(secretreq).getData());
```
Comment (seeflood) [inline examples-secret/src/main/java/io/mosn/layotto/examples/secret/Secret.java:28]: Could u add some assertions here ?
For example, in `src/main/java/io/mosn/layotto/examples/lock/Lock.java`, we assert that the response is true/false:
image
The benefit of these assertions is that we can run these demos as test cases. 
For more details, check https://mosn.io/layotto/#/zh/development/test-quickstart?id=q-%e5%a6%82%e4%bd%95%e7%bc%96%e5%86%99%e5%85%b7%e6%9c%89quot%e5%8f%af%e6%b5%8b%e8%af%95%e6%80%a7quot%e7%9a%84%e6%96%87%e6%a1%a3

**Review by seeflood — CHANGES_REQUESTED** [review]

File: examples-secret/src/main/java/io/mosn/layotto/examples/secret/Secret.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+            /*GetBulkSecret*/
+            Map<String, String> meta2 = new HashMap<>();
+            GetBulkSecretRequest bulksecretreq = new GetBulkSecretRequest();
+            bulksecretreq.setStoreName(storeName);
+            bulksecretreq.setMetaData(meta2);
+            System.out.println(client.getBulkSecret(bulksecretreq).getData());
```
Comment (seeflood) [inline examples-secret/src/main/java/io/mosn/layotto/examples/secret/Secret.java:35]: same as above

**Review by seeflood — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/spec/sdk/runtime/v1/domain/secret/GetBulkSecretRequest.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    public void setStoreName(String storeName) {
+        this.storeName = storeName;
+    }
+
+    public Map<String, String> getMetaData() {
+        if (metaData == null) {
+            metaData = new HashMap<>();
+        }
```
Comment (seeflood) [inline sdk/src/main/java/spec/sdk/runtime/v1/domain/secret/GetBulkSecretRequest.java:22]: getter methods shouldn't modify the value.
I prefer to `return metaData` directly.

**Review by seeflood — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/io/mosn/layotto/v1/RuntimeClientGrpc.java [hunk 9]
```diff
@@ -1101,4 +1105,53 @@ public UnlockResponse unlock(UnlockRequest request) {
...
 
     }
+
+    @Override
+    public  GetSecretResponse getSecret(GetSecretRequest req){
+        try {
+            RuntimeProto.GetSecretRequest request = RuntimeProto.GetSecretRequest.newBuilder()
+                    .setStoreName(req.getStoreName())
+                    .setKey(req.getKey())
+                    .putAllMetadata(req.getMetaData())
```
Comment (seeflood) [inline sdk/src/main/java/io/mosn/layotto/v1/RuntimeClientGrpc.java:1115]: Oh no, we have to check if the `req.getMetaData()` returns null here 😢

File: sdk/src/main/java/io/mosn/layotto/v1/RuntimeClientGrpc.java [hunk 10]
```diff
@@ -1101,4 +1105,53 @@ public UnlockResponse unlock(UnlockRequest request) {
...
+            throw new RuntimeClientException(e);
+        }
+
+    }
+    @Override
+    public  GetBulkSecretResponse getBulkSecret(GetBulkSecretRequest req){
+        try {
+            RuntimeProto.GetBulkSecretRequest request = RuntimeProto.GetBulkSecretRequest.newBuilder()
+                    .setStoreName(req.getStoreName())
+                    .putAllMetadata(req.getMetaData())
```
Comment (seeflood) [inline sdk/src/main/java/io/mosn/layotto/v1/RuntimeClientGrpc.java:1134]: Same as above

**Review by seeflood — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/io/mosn/layotto/v1/RuntimeClientGrpc.java [hunk 11]
```diff
@@ -1101,4 +1105,53 @@ public UnlockResponse unlock(UnlockRequest request) {
...
+                    .build();
+
+            RuntimeProto.GetBulkSecretResponse response = runtimeStubManager.getBlockingStub()
+                    .getBulkSecret(request);
+            GetBulkSecretResponse getBulkSecretResponse = new GetBulkSecretResponse();
+            Map<String, Map<String, String>> tempMap = new HashMap<String,Map<String,String>>();
+            for(Map.Entry<String,RuntimeProto.SecretResponse> entry : response.getDataMap().entrySet()){
```
Comment (seeflood) [inline sdk/src/main/java/io/mosn/layotto/v1/RuntimeClientGrpc.java:1141]: Maybe we should check if `response.getDataMap()`  returns null here?

**Review by seeflood — CHANGES_REQUESTED** [review]

File: examples-secret/src/main/java/io/mosn/layotto/examples/secret/Secret.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            GetSecretRequest secretreq = new GetSecretRequest();
+            secretreq.setStoreName(storeName);
+            secretreq.setKey(key);
+            secretreq.setMetaData(meta);
+            GetSecretResponse response = client.getSecret(secretreq);
+            System.out.println(response.getData());
+            assertMap(response.getData(),key,value);
+
+            /*GetBulkSecret*/
+            Map<String, String> meta2 = new HashMap<>();
```
Comment (seeflood) [inline examples-secret/src/main/java/io/mosn/layotto/examples/secret/Secret.java:38]: By the way, to run this demo in Layotto CI, you need to modify the [secret API quickstart documentation](https://mosn.io/layotto/#/zh/start/secret/start) .
image

You can take https://github.com/mosn/layotto/pull/783 as an example.
For more details, read https://mosn.io/layotto/#/zh/development/test-quickstart

**Review by seeflood — CHANGES_REQUESTED** [review]

**Review by seeflood — CHANGES_REQUESTED** [review]

**Comment by fft0518** [comment]
image
@seeflood  How can I resolve conflicts？

**Review by seeflood — APPROVED** [review]
lgtm. Thanks

**Comment by fft0518** [comment]
> lgtm. Thanks

Thanks!

**Pull request merged**